### ISSUE TITLE
4002: FPI restrict opener -> active

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#867]
originally changed in #840 

change my mind :)

active
- pros:
- cons:

inactive
- pros:
- cons:

fill in, or I'll get to it later: also see conversation in https://github.com/ghacksuserjs/ghacks-user.js/commit/5672bc8cc8a978eed39e39b009cb909882ce4db0

**Comment by Thorin-Oakenpants** [comment]
I have better things to do than argue over this, but I would like to follow best practice (as I see it)

Enforcing the default here carries risk (albeit super highly unlikely to ever happen). The only people who would flip it are the Tor Uplift guys themselves, and they would do so for good reason. This means end users (if we don't pick up on it, or they don't update their user.js) would be stuck with a bad pref value <-- **this** is the reason why I made it inactive

As an added bonus (**but not the reason why**), it reduces the number of modified prefs users have to consider/troubleshoot etc. Whether this helps someone or not is debatable (e.g users can already see the default tag and skip over it, as earthlng pointed out)

I don't really see any risk/downside to keeping it inactive, but a super-uber-tiny chance that making it active does. If you can provide a **compelling** argument for it to be active, then reopen and state the case (e.g. people getting bad advise to mess with the default is about the only reason I can think of, but anecdotally I haven't seen that)

---

Read if you want to....

The user.js tries to follow some "rules" for consistency, but not documented or even perceived as the same (we have our own styles), as earthlng and myself have crafted it. e.g. [we try to] use the word `enforce` in descriptions for default values, [we try to] add warning tags only for inactive prefs (so that these are limited and stand out) and [setup] tags for active prefs (less items to check for new users), and so on. It's not perfect, and due to the fluid nature of pref values and the pref's intent, etc, I can already spot some "breakage" in those "rules". And there will propabably always be exceptions to the "rule of thumb" in how we perceive this

earthlng [said](https://github.com/ghacksuserjs/ghacks-user.js/commit/5672bc8cc8a978eed39e39b009cb909882ce4db0#commitcomment-36466078)
> MO inactive prefs are inactive for 1 of 2 reasons: to allow users to either harden or relax our template. Adding a 3rd kind where the value matches FF's default is just confusing.

I've never really seen that pattern in our "rules", because it's not a rule, but a side-effect: any pref that is inactive -> active creates a change which logically hardens or relaxes (depending on the end user's perspective/threat model). We already have other examples of the "3rd kind" (see 1032 for example which I suggested removing but didn't deep dive in the end, and 1600s which we need for the info factor, but we could change the values just as easily). It's a small set, and the pattern of 1 of 2 reasons is a byproduct of the nature of the user.js IMO.

But there isn't really a third kind per se (certainly not something I've seen as a pattern). Sure, we could slice and dice all inactive prefs into sub-categories: such as inactive-at- default and inactive-not-at-default and subconsciously apply some sort of "rule" on how we frame it. But I don't see the point. At this level it's more about what the pref does and any risk assessment. We could enforce such a rule: e.g. we remove 1032 as a waste of space, we change the values used in the 1600s, and make 1402 active. But it's not a rule that makes sense to me

Lets have a look at a smaller slightly different example (how I see it, without deep diving too much), which includes 1032, 1600s, and 1402

inactive items with `[default`
```
- 1032: favicon in web notifications - kept in case Mozilla flip it? I tried to deep dive it
- 1600s: six referer prefs - definitely kept for the info factor, defaults used in the user_pref because we don't really have a position on those: end user choice or as we said, use an extension
- 4002: FPI post message and now this issue: FPI restrict opener
```

active items with  `[default` (enforcing)
```
- 0401: blocklists - for safety due to lots of bad advice out there
- 0602: prefetching - it's not default in ESR
- 0606: pings - because Mozilla will flip it to follow the w3c guidelines or whatever: same as chrome
- 1240: mixed content active
- 1273: display insecure icon - it;s not default in ESR
- 2200s: three window_open_features items - included to complete the set
- 2403: middlemouse.paste - diff defaults per OS
- 2621: pdfjs - debatable, but read the description
- 2660: extension scopes -at least one of them at default. Kept as a pair since they act together
- 2680: CSP - because it's a really important security thing and also kept for the knowledge factor (as pointed out by earthlng when recent revisited)
- 2720: IDB - because it breaks extensions, relic from pre FF63 days, is about to be derpecated
```

I don't see any logic here that these are this way because they harden/relax. It's a small sample, and overall there isn't much that would harden (active prefs outnumber inactive: and by default we harden way more than relax). What I do see is that each pref is evaluated on what it does, how people use it and/or to correct previous bad advice, how it might change in the future, etc. Each pref is assessed on it's own merits, and if it can fit a perceived "rule", cool: if not, so be it.

I would see this same pattern if I were to analyze all inactive prefs - i.e it is the intent of the pref (and most (all?) are configured to harden if active) and the risk it carries and why we included it in the first place that dictates it's state - not because it hardens or relaxes, that's just a byproduct and they all do that.

**Comment by Thorin-Oakenpants** [comment]
> I have better things to do than argue over this

Just FYI

You know, sometimes, I wish people would just trust in me a bit more rather than arguing things to death and not listening (apologies but that still fucking pisses me off - you know what I'm talking about: hence why I'm so aloof and zero fucks given anymore - maybe my mood will wear off in a year or two). Here is an email I received almost two years ago

![email](https://user-images.githubusercontent.com/16656956/74369402-3ae5e200-4dcd-11ea-9a82-75a8a633b5be.png)

The chances of the pref getting flipped is exactly `ZERO`, and (there's a typo: ever should be never? "Except it was ever actually used"), it was never used by Tor Browser (I guess they got busy elsewhere and eventually the ESR60 rolled around)

It's a little hard to follow what Tom said: if it was added for TB for backwards compat, then why did it have an affect in Firefox (one edge case mentioned: which must be Firefox since TB never added the patch) - confused - @tomrittervg (if you have time, otherwise, save yourself for better things: this is not important)

**Comment by tomrittervg** [comment]
1. Despite sending that email, I never actually removed it. The patch I was working on got sidelined so while I feel able to remove it anytime I want without telling anyone, I haven't.
2. It was added to be compatible with Tor Browser's existing behavior, but they never turned it on. So the FPI behavior they got from Firefox was slightly more restrictive (safer) than their old behavior.
3. We did identify one login flow that Firefox's FPI broke but Tor Browser's old behavior (the behavior you get if you disable that pref) allowed.

I see no reason to ever disable this pref unless you're testing FPI login flows and want to see if this fixes a particular one.
